File: examples/pythoncommerce/commerce/service.py
"""Application service with conservative static call evidence."""

from collections.abc import Awaitable, Callable

from commerce.models import Order
from commerce.repository import OrderRepository


def traced(function: Callable[..., Awaitable[None]]) -> Callable[..., Awaitable[None]]:
    """Decorator syntax is indexed; its runtime transformation stays unknown."""
    return function


def validate_order(order: Order) -> None:
    if not order.id:
        raise ValueError("order id is required")


class OrderService:
    def __init__(self, repository: OrderRepository) -> None:
        self._repository = repository

    @traced
    async def submit(self, order: Order) -> None:
        validate_order(order)
        await self._repository.save(order)

File: examples/rustcommerce/src/service.rs
use crate::models::Order;
use crate::repository::Repository as OrderRepository;

/// OrderService coordinates creation and persistence of orders.
pub struct OrderService<R: OrderRepository> {
    repository: R,
}

impl<R: OrderRepository> OrderService<R> {
    pub fn new(repository: R) -> Self {
        Self { repository }
    }

    pub async fn submit(&mut self, id: &str, customer_id: &str) -> Order {
        let order = build_order(id, customer_id);
        self.repository.save(order.clone());
        order
    }

    pub fn find(&self, id: &str) -> Option<&Order> {
        self.repository.find(id)
    }
}

fn build_order(id: &str, customer_id: &str) -> Order {
    Order::new(id, customer_id)
}
